esp-hal/src/sdm.rs

```diff
@@ -0,0 +1,257 @@
+//! Sigma-Delta modulation peripheral driver.
+
+use fugit::HertzU32;
+
+use crate::{
+    clock::Clocks,
+    gpio::{OutputSignal, PeripheralOutput},
+    peripheral::{Peripheral, PeripheralRef},
+    peripherals,
+    private,
+};
+
+/// Sigma-Delta modulation peripheral driver.
+pub struct Sdm<'d, SD> {
+    _sd: PeripheralRef<'d, SD>,
+    channels_usage: u8,
+}
+
+/// Channel errors
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// Prescale out of range
+    PrescaleRange,
+    /// No free channels to use
+    NoChannels,
+}
+
+impl<'d, SD> Sdm<'d, SD>
+where
+    SD: RegisterAccess,
+{
+    /// Initialize driver using a given SD instance.
+    pub fn new(sd_instance: impl crate::peripheral::Peripheral<P = SD> + 'd) -> Self {
+        Self {
+            _sd: sd_instance.into_ref(),
+            channels_usage: 0,
+        }
+    }
+
+    /// Configure and acquire channel.
+    pub fn enable_pin<PIN: PeripheralOutput>(
+        &mut self,
+        pin: impl Peripheral<P = PIN> + 'd,
+        frequency: HertzU32,
+    ) -> Result<Channel<'_, SD, PIN>, Error> {
+        crate::into_ref!(pin);
+
+        let chs = self.channels_usage;
+        let chidx = self.alloc_channel()?;
+        let signal = CHANNELS[chidx as usize];
+
+        if chs == 0 {
+            SD::enable_clock(true);
+        }
+
+        pin.connect_peripheral_to_output(signal, private::Internal);
+
+        let mut channel = Channel {
+            _sdm: self,
+            pin,
+            chidx,
+        };
+
+        channel.set_frequency(frequency)?;
+
+        Ok(channel)
+    }
+
+    /// Deconfigure and release channel.
+    pub fn disable_pin<PIN: PeripheralOutput>(&mut self, channel: Channel<'d, SD, PIN>) {
+        let Channel { mut pin, chidx, .. } = channel;
+
+        let signal = CHANNELS[chidx as usize];
+
+        pin.disconnect_from_peripheral_output(signal, private::Internal);
+
+        self.dealloc_channel(chidx);
+
+        if self.channels_usage == 0 {
+            SD::enable_clock(false);
+        }
+    }
+
+    fn alloc_channel(&mut self) -> Result<u8, Error> {
```

Comment:
This is actually pretty annoying in both Pcnt and SysTimer :)

Reply:
How so? (is it something type erasure can solve?)

Reply:
> (is it something type erasure can solve?)

No, it's manual vs automatic resource management. There are pros and cons to each approach (not compiling vs simplicity), and I guess so far we've been leaning towards manual management pretty much everywhere. It's just... well, you don't *need* to manually track which DMA channels you've used where, or which timer is running behind a PWM. If two resources are exactly equivalent, why make the users job harder?

To be fair, since we're using manual management everywhere, we might as well just stick to it here as well. It's just somewhat of a chore to deal with :)

Reply:
Yeah that makes sense, I mostly like things manual because there's less error handling involved. It's nice to be able to statically prove that I have enough timers or channels for my application.

> If two resources are exactly equivalent, why make the users job harder?

I guess this is getting into the erasure side of things. I'm fully on board as long as it's optional. Sometimes the specific type is handy when writing interrupt handlers.

Reply:
Seems pcnt driver doesn't implement disabling channels and disconnecting inputs from signals at all. I think this is necessary feature and should be implemented in any case.

Reply:
Disconnecting inputs isn't strictly necessary, since it's an input. In this case we're dealing with an output pin so it's somewhat necessary. I say somewhat because other drivers don't really do this.

Though this is orthogonal to manual resource management.